haskell/test/SMT/Model/Types.hs

```diff
@@ -0,0 +1,117 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module SMT.Model.Types
+  ( ModelT (..)
+  , ModelState (..)
+  , runModelT
+  , getStore
+  , getConfig
+  , modifyStore
+
+  , ModelError (..)
+  , contractErrorToModelError
+  )where
+
+import Universum
+
+import Fmt
+import Lorentz hiding (not, cast, get)
+import Control.Monad.Except (MonadError)
+
+import Ligo.BaseDAO.Types
+
+-- | Transformer used in haskell implementation of BaseDAO
+newtype ModelT a = ModelT
+  { unModelT :: (ExceptT ModelError $ State ModelState) a
+  } deriving newtype (Functor, Applicative, Monad, MonadState ModelState, MonadError ModelError)
+
+-- | A type that keep track of blockchain state
+data ModelState = ModelState
+  { msFullStorage :: FullStorage
+  , msLevel :: Natural
+  , msChainId :: ChainId
+  , msSelfAddress :: Address
+  , msGovernanceTokenAddress :: Address
+  } deriving stock (Eq, Generic, Show)
+
+runModelT :: ModelT a -> ModelState -> (Either ModelError a, ModelState)
```

Comment:
Hmm, I think we should separate them, since even when the result is `ModelError`, we still want to check if `ModelState` is stay the same. What do you think about this?

Reply:
I don't think so, because in case of a failure in Michelson the whole operation gets reverted, so the `ModelState` should not be important AFAIK.

Reply:
Since I put the current `level` as part of `ModelState`, we have to keep track of it. So even in case of error, we still need to return the `ModelState`. Unless we make another type to keep track of the blockchain state.

Reply:
:thinking:  I see, I didn't consider `level`.

However in that case would it not be easier to have a function just to update the level (as you already had before my last suggestion) and call it _before_ running `runModelT`?

This way if an error occurred (and we get a `Left`) we can move to the next step with the previous state (with the level updated) whereas if no error occur we have a new state to use for the next step instead.

Reply:
Make sense, I've made the adjustment. This should be ready for re-review.